File: KalmanFilter/C/KalmanFilter.c
/*
 * Owner: H. Yildiz
 * Brief: KalmanFilter.py ported to C
 * Note: gcc -o KalmanFilter KalmanFilter.c MatrixMath.c -lm
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h> //memcpy()
#include <math.h>   //sin(), cos(), pow()
#include "MatrixMath.h"
#include "KalmanFilter.h"

#define LEN(arr) ((int) (sizeof (arr) / sizeof (arr)[0])) //LEN(arr) = rows, LEN(arr[0]) = columns
#define PI 3.14159265
#define RADIUS_EARTH 6378137 // Radius of the earth is 6378137m

/*
 * Inputs:
 *   t = time interval between itterations
 *   u_k = acceleration
 *   theta = Angle apposed to true North
 *
 * 3 Step process (as seen in the report):
 * Step Prediction (1):
 * 1: X_kp = AX_k-1 + Bu_k + w_k                           // Predict current elements (position/volicity in this example)
 * 2: P_kp = AP_k-1A^T + Q_k                               // Process converiance matrix to calculate the Kalman gain in step 4
 * Step Update (2):
 * 3: K = (P_kpH^T)/(HP_kpH^T + R)                         // Calculate the Kalman Gain
 * 4: Y_k = CY_km + z_m                                    // Convert input to correct form
 * 5: X_k = X_kp + K[Y_k - HX_kp]                          // Update prediction from step 2 with the measurement
 * 6: P_k = (I - KH)P_kp                                   // Calculate new error
 * Step Reload (3):
 * 7: X_k-1 = X_k and P_k-1 = P_k                          // Current becomes previous in the next iteration (current is output)
*/

void prediction(struct KalmanSet *set, float u_k, int theta)
{
    // 1: X_kp = AX_k-1 + Bu_k + w_k
    float A[ROW2][COL2] = {1,0,0,1};
    float B[ROW2][COL1] = {cos(theta * PI/180), sin(theta * PI/180) / cos(set->X_k1[0][0] * PI/180)};
    int w_k = 0;

    // Convert meters to delta latitude
    u_k = (u_k / 3.6) / ((2 * PI/360) * RADIUS_EARTH);

    //self.X_kp = np.dot(A, self.X_k1) + np.dot(B, u_k) + w_k
    float temp[ROW2][COL1];
    float temp2[ROW2][COL1];

    matrix_multiplication(LEN(A), LEN(A[0]), A, LEN(set->X_k1[0]), set->X_k1, temp);
    matrix_dot_vector(LEN(B), LEN(B[0]), B, u_k, temp2);
    matrix_addition(ROW2, COL1, temp, temp2, set->X_kp);

    // 2: P_kp = AP_k-1A^T + Qk
    float temp3[ROW2][COL2];
    float temp4[ROW2][COL2];
    float temp5[ROW2][COL2];

    float Q_k = pow(0.1, 2);

    matrix_multiplication(LEN(A), LEN(A[0]), A, LEN(set->P_k1[0]), set->P_k1, temp3);
    matrix_transpose(LEN(A), LEN(A[0]), A, temp4);
    matrix_multiplication(LEN(set->P_k1), LEN(set->P_k1[0]), set->P_k1, LEN(temp4[0]), temp4, temp5);
    matrix_additionSingle(LEN(temp5), LEN(temp5[0]), temp5, Q_k, set->P_kp);
    matrix_keep_diagonal(LEN(set->P_kp), LEN(set->P_kp[0]), set->P_kp, set->P_kp);
}

void update(struct KalmanSet *set, float Oe, float Mea[][COL1])
{
    // 3: K = (P_kpH^T)/(HP_pkH^T + R) // H = identity matrix
    float temp[ROW2][COL2];
    float temp2[ROW2][COL2];

    float Rdop[ROW2][COL2] = {pow(Oe,2), 0, 0, pow(Oe,2)};

    matrix_addition(LEN(Rdop), LEN(Rdop[0]), Rdop, set->P_kp, temp);
    // Division A/B = A*B^-1, this is also the only place in the code #IND00 error's can occur.
    matrix_inverse(temp, temp2);
    matrix_multiplication(LEN(set->P_kp), LEN(set->P_kp[0]), set->P_kp, LEN(temp2[0]), temp2, set->K_g);

    // 4: Y_k = CY_km + z_m // C = identity matrix and z_m = 0
    matrix_duplicate(LEN(set->Y_k), LEN(set->Y_k[0]), Mea, set->Y_k);

    // 5: X_k = X_kp + K[Y_k - HX_kp] // H = identity matrix
    float temp3[ROW2][COL1];
    float temp4[ROW2][COL1];

    matrix_substraction(LEN(set->Y_k), LEN(set->Y_k[0]), set->Y_k, set->X_kp, temp3);
    matrix_multiplication(LEN(set->K_g), LEN(set->K_g[0]), set->K_g, LEN(temp3[0]), temp3, temp4);
    matrix_addition(LEN(set->X_kp), LEN(set->X_kp[0]), set->X_kp, temp4, set->X_k);

    // 6: P_k = (I - KH)P_kp // I = H = identity matrix
    float identity[ROW2][COL2] = {1,0,0,1};
    matrix_substraction(LEN(identity), LEN(identity[0]), identity, set->K_g, temp);
    matrix_multiplication(LEN(temp), LEN(temp[0]), temp, LEN(set->P_kp[0]), set->P_kp, set->P_k);
}

void reload(struct KalmanSet *set)
{
    // 7: X_k-1 = X_k and P_k-1 = P_k
    matrix_duplicate(LEN(set->X_k1), LEN(set->X_k1[0]), set->X_k, set->X_k1);
    matrix_duplicate(LEN(set->P_k1), LEN(set->P_k1[0]), set->P_k, set->P_k1);
}


// https://stackoverflow.com/questions/12911299/read-csv-file-in-c
const char* getfield(char* line, int num)
{
    const char* tok;
    char *tmp = line;
    for (tok = strtok(tmp, ",");
            tok && *tok;
            tok = strtok(NULL, ",\n"))
    {
        if (!--num)
            return tok;
    }
    return NULL;
}

int main(void)
{
    FILE* input_stream = fopen("../Datasets/kalmansetLondon.csv", "r");
    FILE *output_file = fopen("../Datasets/ksetOutputLondonC.csv", "w");
    
    char line[1024];
    fgets(line, 1024, stream); // Skip first line

    struct KalmanSet package = {.X_k1 = {51.54196,-0.0143366666666667}, .P_k1 = {0.8, 0, 0, 0.8}};
    float timestamp, prevTimestamp;
    float measurements[2][1];
    float hdop;

    int line_number = 1;

    while (fgets(line, 1024, stream))
    {
        // If the second counter goes from 59 to 00, adjust
        timestamp = strtod(getfield(strdup(line), 1), NULL);
        if ((timestamp-prevTimestamp) == 41)
            prevTimestamp += 40;
        else if ((timestamp-prevTimestamp) == 4041)
            prevTimestamp += 4040;

        prediction(&package, (timestamp - prevTimestamp) * strtod(getfield(strdup(line), 6), NULL), strtod(getfield(strdup(line), 4), NULL));

        measurements[0][0] = strtod(getfield(strdup(line), 2), NULL);
        measurements[1][0] = strtod(getfield(strdup(line), 3), NULL);

        hdop = strtod(getfield(strdup(line),5), NULL);
        if (hdop == 0)
            hdop = 1;
        update(&package, hdop, measurements);

        reload(&package);
        fprintf(output_file, "%.13g, %.13g, %.13g, %.13g, %d\n", measurements[0][0], measurements[1][0], package.X_k[0][0], package.X_k[1][0], line_number);

        prevTimestamp = timestamp;

        line_number++;
    }
    fclose(output_file);
    fclose(stream);
}

File: KalmanFilter/Python/KalmanFilter.py
###
# = Kalman-filter
# : H. Yildiz
# /
###

##
#   xInputs:
#   X_k-1 = X_0 = Initial location/velocity
#   P_k-1 = P_0 = Initial error of the Kalman calculation
#   Y_km = Location and velocity measurement
##

##
# 3 Step process (as seen in the report):
# Step Prediction (1):
# 1: X_kp = AX_k-1 + Bu_k + w_k                           // Predict current elements (position/volicity in this example)
# 2: P_kp = AP_k-1A^T + Q_k                               // Process converiance matrix to calculate the Kalman gain in step 4
# Step Update (2):
# 3: K = (P_kpH^T)/(HP_kpH^T + R)                         // Calculate the Kalman Gain
# 4: Y_k = CY_km + z_m                                    // Convert input to correct form
# 5: X_k = X_kp + K[Y_k - HX_kp]                          // Update prediction from step 2 with the measurement
# 6: P_k = (I - KH)P_kp                                   // Calculate new error
# Step Reload (3):
# 7: X_k-1 = X_k and P_k-1 = P_k                          // Current becomes previous in the next iteration (current is output)
##

##
#   xOuputs:
#   X_k = Matrix containing the filtered location and velocity.
#   P_k = Matrix containing the error of the location and velocity
##

from math import sin, cos, pi
import numpy as np

class KalmanFilter:
    # Inputs:
    #   t = time interval between itterations
    #   u_k = acceleration
    #   log = weather or not to log the output
    def __init__(self, X_0, P_0, logger):
        if logger:
            self.logging = True
            self.logger = open("../../logs/kalmanfilterpyIgnore.txt", "w")
            self.logger.write("--- New Run ---")
        else:
            self.logging = False

        self.X_k1 = X_0
        self.P_k1 = np.array([[P_0[0][0]**2, 0],
                              [0, P_0[1][0]**2]])

    def __del__(self):
        if self.logging:
            self.logger.close()

    # Inputs:
    #   t = time interval between itterations
    #   u_k = acceleration
    #   theta = Angle apposed to true North
    def prediction(self, u_k, theta):
        # 1: X_kp = AX_k-1 + Bu_k + w_k
        A = np.array([[1, 0],
                      [0, 1]])
        B = np.array([[cos(theta*pi/180)], # /180 to convert it to radians and /lat to compensate for the earth
                      [sin(theta*pi/180) / cos(self.X_k1[0][0] * pi / 180)]])
        w_k = 0

        if self.logging:
            self.logger.write("u_k: {0}, Theta: {1}, ".format(u_k, theta))

        # Convert meters to delta latitude, radius of the earth is 6378137m
        u_k = (u_k / 3.6) / ((2 * pi / 360) * 6378137)

        self.X_kp = np.dot(A, self.X_k1) + np.dot(B, u_k) + w_k

        if self.logging:
            self.logger.write("B: {0} {1}, X_k1: {2} {3}, u_k2: {4}, ".format(B[0][0], B[1][0], self.X_k1[0][0], self.X_k1[1][0], u_k))

        # 2: P_kp = AP_k-1A^T + Qk
        Q_k = 0.1**2
        self.P_kp = np.dot(np.dot(A, self.P_k1), np.transpose(A)) + Q_k
        self.P_kp = np.diag(np.diag(self.P_kp)) # Only keep diagonal

    # Inputs:
    # Oe = Observational errors, matrix of error in measurement
    # MEA = Location and velocity measurement
    def update(self, Oe, MEA):
        # 3: K = (P_kpH^T)/(HP_pkH^T + R) // H = identity matrix
        R = np.array([[Oe[0][0]**2, 0],
                      [0, Oe[1][0]**2]])

        self.K = np.divide(self.P_kp,(self.P_kp + R))
        self.K = np.nan_to_num(self.K) # Remove nan's from 0/0

        if self.logging:
            self.logger.write("K: {0} {1}".format(self.K[0][0], self.K[1][1]))

        # 4: Y_k = CY_km + z_m // C = identity matrix
        z_m = 0
        self.Y_k = MEA

        # 5: X_k = X_kp + K[Y_k - HX_kp] // H = identity matrix
        self.X_k = self.X_kp + np.dot(self.K, (self.Y_k - self.X_kp))

        # 6: P_k = (I - KH)P_kp // I = H = identity matrix
        self.P_k = np.dot((np.identity(2) - self.K), self.P_kp)


    def reload(self):
        # 7: X_k-1 = X_k and P_k-1 = P_k
        self.X_k1 = self.X_k
        self.P_k1 = self.P_k

    # Outputs:
    # X_kp = Predicted location
    def get_prediction(self):
        return self.X_kp

    # Outputs:
    # X_k = Location and velocity estimate
    def get_output(self):
        return self.X_k

    # Outputs:
    # P_k = Error in estimate
    def get_output_error(self):
        return self.P_k
